x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -224,14 +224,19 @@ private void getUserClaims(@Nullable AccessToken accessToken, JWT idToken, Nonce
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace("Received and validated the Id Token for the user: [{}]", verifiedIdTokenClaims);
             }
+            // Add the Id Token string as a synthetic claim
+            final JSONObject verifiedIdTokenClaimsObject = verifiedIdTokenClaims.toJSONObject();
+            final JWTClaimsSet idTokenClaim = new JWTClaimsSet.Builder().claim("id_token_hint", idToken.serialize()).build();
+            verifiedIdTokenClaimsObject.merge(idTokenClaim.toJSONObject());
+            final JWTClaimsSet enrichedVerifiedIdTokenClaims = JWTClaimsSet.parse(verifiedIdTokenClaimsObject);
```

Comment:
This feels like a hack to me. I would prefer a `ActionListener<Tuple<JWT , JWTClaimsSet>> claimsListener` listener. This also feels a bit hacky, so you could create a container class, and pass that as the type of the listener. If you don't dig this too, you could use the `threadContext`.

From my perspective, this would've been the last alternative I would consider, but as it stands it's not a reason against LGTM.

Reply:
I don't necessarily agree on the order of preference, but this is not a strong opinion. I'm gonna go with using the listener to return the serialized JWT as a String

Reply:
That was a poor initial choice I think. It ends up with passing around the serialized string or the JWT in internal methods all over the OpenIdConnectAuthenticator.java along with the listener. Even after we have consumed the Id token :/ How would you use the `threadContext` for this ?

Reply:
You need to carry the `ThreadPool` in the `OpenIdConnectRealm` and into the `OpenIdConnectAuthenticatior`, and then have `threadPool.getThreadContext().putTransient("_oidc_received_id_token", itToken.serialize())` (untested).

Reply:
I think the threadContext is not a great idea. It can easily be abused and I feel like that would be what this is doing. I wonder if we can pass a Function that takes the verified claims and returns an enriched verified claims?

Reply:
@jaymode Could you slightly elaborate on what you mean with "pass a Function that takes the verified claims and returns an enriched verified claims" ? I'm thinking that if we will "enrich" the JwtClaimsSet eitherway, what are the benefits compared to the way it's implemented now?

I was thinking about this again and the synthetic claim approach doesn't seem hacky after all. At least not hacky enough to justify the listener or the threadContext approach

Reply:
I was thinking having a parameter of type `Function<JWTClaimsSet, JWTClaimsSet> claimsEnricher` that did what this was doing but it’s not necessary. Just a alternative to the other alternatives. I am fine with this as is

Reply:
@albertzaharovits I'm going to merge this as - is to get the Realm effort moving. I'm open to discussing this again though, this is in no way a dismissal of the feedback. My current thinking is that the current implementation is not hacky enough, especially considering the alternatives, but I'd be open to be swayed